gtest/src/manager/block_exec.rs

```diff
@@ -137,10 +139,108 @@ impl ExtManager {
+            .first_incomplete_tasks_block
+            .take()
+            .map(|block| {
+                self.gas_allowance = self
+                    .gas_allowance
+                    .saturating_sub(Gas(db_weights.write.ref_time));
+                (block, false)
+            })
+            .unwrap_or_else(|| {
+                self.gas_allowance = self
+                    .gas_allowance
+                    .saturating_sub(Gas(db_weights.read.ref_time));
+                (current_bn, true)
+            });
+
+        // When we had to stop processing due to insufficient gas allowance.
+        let mut stopped_at = None;
+
+        let missing_blocks = first_incomplete_block..=current_bn;
+        for bn in missing_blocks {
+            if self.gas_allowance.0 <= db_weights.write.ref_time.saturating_mul(2) {
+                stopped_at = Some(bn);
+                log::debug!(
+                    "Stopped processing tasks at: {stopped_at:?} due to insufficient allowance"
+                );
+                break;
+            }
+
+            let mut last_task = None;
+            for task in self.task_pool.drain_prefix_keys(bn) {
+                // decreasing allowance due to DB deletion
+                self.on_task_pool_change();
+
+                let max_task_gas = get_maximum_task_gas(&task);
+                log::debug!(
+                    "⚙️  Processing task {task:?} at the block {bn}, max gas = {max_task_gas}"
+                );
+
+                if self.gas_allowance.saturating_sub(max_task_gas) <= Gas(db_weights.write.ref_time)
+                {
+                    // Since the task is not processed write DB cost should be refunded.
+                    // In the same time gas allowance should be charged for read DB cost.
+                    self.gas_allowance = self
+                        .gas_allowance
+                        .saturating_add(Gas(db_weights.write.ref_time))
+                        .saturating_sub(Gas(db_weights.read.ref_time));
+
+                    last_task = Some(task);
+
+                    log::debug!("Not enough gas to process task at {bn:?}");
+
+                    break;
+                }
+
+                let task_gas = task.process_with(self);
+
+                self.gas_allowance = self.gas_allowance.saturating_sub(Gas(task_gas));
+
+                if self.gas_allowance <= Gas(db_weights.write.ref_time + db_weights.read.ref_time) {
+                    stopped_at = Some(bn);
+                    log::debug!("Stopping processing tasks at (read next): {stopped_at:?}");
+                    break;
+                }
+            }
+
+            if let Some(task) = last_task {
+                stopped_at = Some(bn);
+
+                self.gas_allowance = self
+                    .gas_allowance
+                    .saturating_add(Gas(db_weights.write.ref_time));
+
+                self.task_pool.add(bn, task.clone()).unwrap_or_else(|e| {
```

Comment:
Seems no need for this
```suggestion
                self.task_pool.add(bn, task).unwrap_or_else(|e| {
```